### ISSUE TITLE
Unspecified Behavior for Callee Leaving During CALL

### ISSUE THREAD

**Issue by ecorm** [wamp-proto/wamp-proto#476]
This is a continuation of the long discussion in #448, but will focus purely on the behavior of a Callee leaving during a CALL request after the Dealer already sent the INVOCATION (either normal or progressive). The behavior for a Caller leaving during a normal call is underspecified as well, but I prefer to discuss that in a separate issue. I've learned the hard way that that's is easier to introduce small, incremental changes at a time. :grin:

By the end of the discussion in #448, it seemed we were leaning towards the Dealer returning an ERROR message with the `wamp.error.cancelled` URI. This is the current behavior for Crossbar. The other alternative was returning a more-specific URI such as `wamp.error.callee_left`.

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee

    Caller->>Dealer: CALL
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: GOODBYE
    Dealer->>Caller: CALL ERROR (cancelled)
```

If the Caller never sent a CANCEL message, then it can deduce that the `wamp.error.cancelled` was caused by the Callee leaving.

If the Caller did send a CANCEL message, it cannot know if the cancellation was due to normal cancellation, or due to the Callee leaving (unless ERROR payload args are used to convey more information). In either case, the desired effect to cancel the call is the same from the point of view of the Caller.

I propose that the following text be added to the _Remote Procedures Calls_ section of the basic profile:

> ## 6.3 Abnormal Situations
>
> ### 6.3.1 Callee Leaving Before Call Completion
>
> After sending an `INVOCATION` message, if a *Dealer* detects that the *Callee* has terminated its session without sending a `YIELD` or `ERROR` response, then the *Dealer* MUST return an `ERROR` message back to the *Caller* with a `wamp.error.cancelled` URI. The *Dealer* MAY provide additional information via the `ERROR` payload arguments to clarify that the cancellation is due to the *Callee* leaving before the call could be completed.
>
> (insert above sequence diagram)

Another subsection should be added for Caller leaving, but I prefer to create a separate issue to discuss that.

Furthermore, the `wamp.error.cancelled` ERROR URI would be moved from the Advanced profile to the Basic profile.

**Comment by ecorm** [comment]
Pinging @oberstet , @KSDaemon , @Jopie64 who participated in the prior discussion.

**Comment by KSDaemon** [comment]
Thnx for pinging! I'll look on days!

**Comment by KSDaemon** [comment]
Read the proposal. For me, it looks fine and clear.

Rgd q about `wamp.error.cancelled` VS `wamp.error.callee_left` — can't choose the one, both look good :)
From one side: how often does the caller care about the reason why the call failed? I mean from the end-user point of you. If it is really `callee gone` — does it make sense for the user? I don't think so. In most cases, you as a developer will cover this with something general like `an error happens on the server side`. But users are not developers. So for the client implementation developer's side, it may be interesting to have different errors for different cases. IDK...

In general, if we can send different errors — that is good. Protocol provides precise information — how to treat it — is up to the developer.

**Comment by oberstet** [comment]
@ecorm awesome, thanks for hard work and contributing to improve/polish the spec and fill gaps in the spec. also, big +1 for "introduce small, incremental changes at a time" - I think that very much this "last phase" stage of spec work: fill in gaps that related to detailed behavior like this, once there are already multiple implementations.

rgd your proposal for 6.3.1: +1 from my side as well!

the only thing we might want to discuss more in detail is the error URI (since "canceled" is somehow wrong? as no one actively canceled ..). hence, here is a list of URIs we might want to consider

* `wamp.error.canceled`
* `wamp.error.no_such_procedure`
* `wamp.error.no_eligible_callee`
* `wamp.error.runtime_error`
* ???

**Comment by oberstet** [comment]
@ecorm one more detail: what is your take on "cancelled" vs "canceled"?

I note that AB/CB uses both;) I _think_ I once started with 1 l but then did not really watch for it ... well. now, here is what I found .. but hey, I have no clue, not a native speaker: https://www.merriam-webster.com/words-at-play/canceled-or-cancelled

```
oberstet@intel-nuci7:~$ find scm/crossbario/crossbar/crossbar -name "*.py" -exec grep -Hi "canceled" {} \;
scm/crossbario/crossbar/crossbar/bridge/mqtt/test/test_tx.py:    def test_keepalive_canceled_on_lost_connection(self):
scm/crossbario/crossbar/crossbar/node/node.py:                            if e.error != 'wamp.error.canceled':
scm/crossbario/crossbar/crossbar/router/test/test_dealer.py:        self.assertEqual('wamp.error.canceled', msg.error)
scm/crossbario/crossbar/crossbar/router/test/test_dealer.py:        self.assertEqual('wamp.error.canceled', call_error_msg.error)
scm/crossbario/crossbar/crossbar/router/test/test_dealer.py:            session0, message.Error(message.Invocation.MESSAGE_TYPE, invocation_msg0.request, 'wamp.error.canceled'))
scm/crossbario/crossbar/crossbar/router/dealer.py:        'canceled',
scm/crossbario/crossbar/crossbar/router/dealer.py:        self.canceled = False
scm/crossbario/crossbar/crossbar/router/dealer.py:                if invoke.canceled:
scm/crossbario/crossbar/crossbar/router/dealer.py:                    ApplicationError.CANCELED,
scm/crossbario/crossbar/crossbar/router/dealer.py:                            ApplicationError.CANCELED,
scm/crossbario/crossbar/crossbar/router/dealer.py:            if invocation_request.canceled:
scm/crossbario/crossbar/crossbar/router/dealer.py:            invocation_request.canceled = True
scm/crossbario/crossbar/crossbar/router/dealer.py:                    ApplicationError.CANCELED,
scm/crossbario/crossbar/crossbar/router/dealer.py:            if invocation_request.canceled:
scm/crossbario/crossbar/crossbar/router/dealer.py:            else:  # not canceled
scm/crossbario/crossbar/crossbar/router/dealer.py:            if invocation_request.canceled:
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-python/autobahn -name "*.py" -exec grep -Hi "canceled" {} \;
scm/crossbario/autobahn-python/autobahn/websocket/protocol.py:        # timers, which might get set up later, and remembered here to get canceled
scm/crossbario/autobahn-python/autobahn/wamp/exception.py:    CANCELED = "wamp.error.canceled"
scm/crossbario/autobahn-python/autobahn/wamp/exception.py:    A Dealer or Callee canceled a call previously issued (WAMP AP).
scm/crossbario/autobahn-python/autobahn/wamp/message.py:            triggered by the caller (``"wamp.error.canceled"`` - ApplicationError.CANCELED) or
scm/crossbario/autobahn-python/autobahn/wamp/types.py:        :param timeout: Time in seconds after which the call should be automatically canceled.
scm/crossbario/autobahn-python/autobahn/wamp/interfaces.py:        be canceled by canceling the returned Deferred/Future.
oberstet@intel-nuci7:~$ find scm/crossbario/crossbar/crossbar -name "*.py" -exec grep -Hi "cancelled" {} \;
scm/crossbario/crossbar/crossbar/bridge/mqtt/test/test_tx.py:        self.assertTrue(timeout.cancelled)
scm/crossbario/crossbar/crossbar/router/dealer.py:            # we've maybe cancelled this invocation
scm/crossbario/crossbar/crossbar/router/dealer.py:            # we've maybe cancelled this invocation
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-python/autobahn -name "*.py" -exec grep -Hi "cancelled" {} \;
scm/crossbario/autobahn-python/autobahn/twisted/rawsocket.py:from twisted.internet.defer import CancelledError
scm/crossbario/autobahn-python/autobahn/twisted/rawsocket.py:        except CancelledError as e:
scm/crossbario/autobahn-python/autobahn/twisted/rawsocket.py:            self.log.debug("{klass}.stringReceived: WAMP CancelledError - connection will continue!\n{err}",
scm/crossbario/autobahn-python/autobahn/asyncio/component.py:            except asyncio.CancelledError:
scm/crossbario/autobahn-python/autobahn/asyncio/component.py:                log.debug("All task cancelled")
scm/crossbario/autobahn-python/autobahn/asyncio/component.py:        except asyncio.CancelledError:
scm/crossbario/autobahn-python/autobahn/websocket/test/test_websocket_frame.py:                # save it for later (to check it got cancelled)
scm/crossbario/autobahn-python/autobahn/websocket/test/test_websocket_frame.py:                # which should have cancelled the call
scm/crossbario/autobahn-python/autobahn/websocket/test/test_websocket_frame.py:                self.assertTrue(timeout_call.cancelled)
scm/crossbario/autobahn-python/autobahn/wamp/test/test_wamp_protocol.py:            # This could happen if the task waiting on a request gets cancelled
scm/crossbario/autobahn-python/autobahn/wamp/protocol.py:                    # this will result in a CancelledError which will
scm/crossbario/autobahn-python/autobahn/wamp/component.py:                # subclasses to see if this is CancelledError (for
scm/crossbario/autobahn-python/autobahn/wamp/component.py:                # Can connect_f ever be in a cancelled state?
scm/crossbario/autobahn-python/autobahn/wamp/component.py:                # if txaio.using_asyncio and isinstance(fail.value, asyncio.CancelledError):
oberstet@intel-nuci7:~$ 
```

**Comment by ecorm** [comment]
> the only thing we might want to discuss more in detail is the error URI (since "canceled" is somehow wrong? as no one actively canceled

"Canceled" could be semantically correct from the point of view of the call operation: the call was indeed _canceled_, but due to the callee leaving. An `ERROR` payload argument can further explain the cause of the cancellation: explicitly by the caller, or by the callee leaving.

> wamp.error.no_such_procedure

This may not be true in the case of shared registrations. Other callees may still be registered to the same procedure URI.

> wamp.error.no_eligible_callee

This may also not be true for shared registrations. Other callees may have been eligible. Is it expected for the router to attempt other shared registrations if the chosen callee leaves? If so, and the router handles a callee leaving the same as a callee responding with `wamp.error.unavailable`, then this could work.

> wamp.error.runtime_error

I can't find this one in the spec. I'm guessing it's used by Crossbar/Autobahn.

> one more detail: what is your take on "cancelled" vs "canceled"

As the article you cited says, one is the preferred British spelling and the other is the preferred American spelling. The preferred Canadian spelling is usually the British one, but Canadian programmers often use the American spelling for technical terms due to the undeniable US influence on technology.

The spec currently uses "canceled" with a single L, so we should stick to the existing spelling. However, I think I'm going to add "cancelled" as an alternate form in my error URI parser.

**Comment by Jopie64** [comment]
> The behavior for a Caller leaving during a normal call is underspecified as well, but I prefer to discuss that in a separate issue. I've learned the hard way that that's is easier to introduce small, incremental changes at a time.

Although I generally agree with applying small incremental changes, I actually think these cases are so similar that they should be handled together. And if not, we should at least design one with the other in mind.
And since this case is even underspecified for the normal case, I think this should be documented first.

---

As far as I can tell, what we concluded in the other thread already, we're left with 2 options:
* wamp.error.canceled
* wamp.error.callee_left

The pros of the first:
* This error reason is also used as a result of a CANCEL message from the caller. No need to specify another error code
* Seen from the caller the effect is the same: the call was canceled. So caller doesn't have to specify logic for both cases.
* This error code is already used by Crossbar in that case, so Crossbar doesn't have to change
* In most cases callers don't need to add extra logic after this specification change

Cons:
* In a specific race, specified [here](https://github.com/wamp-proto/wamp-proto/issues/448#issuecomment-1449002686), caller can't tell the cancelation reason apart even if it wants to. But again I couldn't think of a reason why it would matter for the caller that a call was cancelled due to callee leaving while it just sent a CANCEL message... The call was canceled, and that's what it wanted.

(Do you notice my bias here? 😁)

Pros and cons for the second one are the reverse 😜

**Comment by ecorm** [comment]
> I actually think these cases are so similar that they should be handled together. And if not, we should at least design one with the other in mind.

After we decide on this issue, we can open another one for the caller leaving scenario. Once there's consensus on both, one of us can submit a single PR that fixes both.

As for the caller leaving error URI, I care more that one be chosen than I care about which specific one is chosen (as long as its reasonable).

**Comment by Jopie64** [comment]
Woops, I wrongly read _callee_ leaving in the normal case. Indeed, _caller_ leaving is something different.

But still, callee leaving should be specified for the non-progressive (or normal) case first i think, and would probably be the same as in the progressive case.

**Comment by ecorm** [comment]
>  wamp.error.no_eligible_callee

> This may also not be true for shared registrations. Other callees may have been eligible. Is it expected for the router to attempt other shared registrations if the chosen callee leaves? If so, and the router handles a callee leaving the same as a callee responding with wamp.error.unavailable, then this could work.

In the case of shared registrations + progressive calls, the dealer is now supposed to stick with the same callee for the duration of the call request. So returning an `wamp.error.no_eligible_callee` ERROR would be a "lie", because there may indeed be other eligible callees. So that rules out `wamp.error.no_eligible_callee` if we want the callee leaving behavior to be the same for all call modes.

I think we're now down to two choices for the ERROR URI returned when the callee leaves after an INVOCATION was sent:

1. `wamp.error.canceled`
2. `wamp.error.callee_left`

Jopie64 has already listed the pros/cons for each above. What is your preferred alternative?

**Comment by ecorm** [comment]
Here's a tally of the "votes" so far for the error URI:

- @ecorm Doesn't have a preference
- @KSDaemon Doesn't seem to have a preference
- @Jopie64 Leans towards `wamp.error.canceled`
- @oberstet Votes for `wamp.error.callee_left`

@oberstet You may have to set your foot down and decide for us. I can do the work in submitting the PR, but I need a decision on the error URI. Before merging the PR, we can wait to bit to see if there are any objections, especially from users who may be affected by a possible change in existing router behavior.

**Comment by oberstet** [comment]
ok, if @Jopie64 is for `wamp.error.canceled`, I change my vote to that as well, so that we have +2 for `wamp.error.canceled`. does that work for you @ecorm and @Jopie64 and all?

**Comment by ecorm** [comment]
> ok, if @Jopie64 is for wamp.error.canceled, I change my vote to that as well, so that we have +2 for wamp.error.canceled. does that work for you @ecorm and @Jopie64 and all?

That works for me, and it has the advantage of not breaking Crossbar and other routers that emulate its behavior. It also avoids introducing a new error URI. As I've already mentioned, an ERROR argument can be used to clarify the origin of the call cancellation.

**Comment by Jopie64** [comment]
Although I do think you raise a good point in favor of `wamp.error.callee_left`, `canceled` still works for me. It's about weighing the pros and cons, and for me it's important that at least every argument is taken into consideration and contributes to the weighing 😊

We can't predict the future so we can just hope we made a good choice.

And now I just thought of another argument in favor of `canceled`: it's easier to change our minds (and change the spec) in the future when we didn't introduce a new error code. Dropping an error code when it's already widely used is harder than introducing it later anyway.

**Comment by ecorm** [comment]
> And now I just thought of another argument in favor of canceled: it's easier to change our minds (and change the spec) in the future when we didn't introduce a new error code. Dropping an error code when it's already widely used is harder than introducing it later anyway.

We'll be essentially documenting a _de facto_ behavior so that it becomes _de jure_. If someone later wants to change that behavior in the standard, they'll have to weigh it against the cost of breaking existing implementations.

**Comment by Jopie64** [comment]
> We'll be essentially documenting a _de facto_ standard behavior so that it becomes _de jure_. If someone later wants to change that behavior in the standard, they'll have to weigh it against the cost of breaking existing implementations.

Agreed. Saying "it's easier" doesn't mean it's easy 😊
I think it's just harder to remove something later than to add something, even when both actions might break compatibility. A removed error code will probably linger around for a long time in client code because old dealer implementations might still send it.

**Comment by KSDaemon** [comment]
Great that we have agreed and solved another piece of spec! 👍🏻

**Comment by ecorm** [comment]
@oberstet @Jopie64 @KSDaemon please see #477 concerning the **caller** leaving during a CALL. I will submit a PR addressing both this issue and #477 once we achieve consensus on the latter.
